Embed for DI

gfiorav opened the pull request:
Fixes #6570 

cc/ @xavijam

xavijam commented:
I feel @juanignaciosl should review this.
And @viddo, could you check frontend part?

gfiorav commented:
Wait! We're not done here! Missing the password protected embeds!

xavijam commented:
Ahhhh, sorry then! my bad :).

viddo commented:
@gfiorav does that affect the frontend parts though?
@xavijam frontend (so far) looks :+1:, but perhaps we should add TrackJS?

gfiorav commented:
@xavijam no problem :joy: 

I'm adding one view for the password form :grin:

xavijam commented:
Could you @gfiorav add the trackjs partial?

gfiorav commented:
Ok, now we're ready for CR. Please @juanignaciosl !

juanignaciosl commented on app/controllers/admin/visualizations_controller.rb:
```diff
@@ -322,7 +322,7 @@ def show_protected_public_map
     return(render_pretty_404) unless @visualization.password_protected? and @visualization.has_password?
 
-    unless @visualization.is_password_valid?(submitted_password)
+    unless @visualization.password_valid?(submitted_password)
```
Brave refactor! :clap:

juanignaciosl commented on app/controllers/carto/api/visualization_presenter.rb:
```diff
@@ -80,7 +80,7 @@ def to_public_poro
           kind:             @visualization.kind,
           privacy:          @visualization.privacy.upcase,
-          likes:            @visualization.likes_count
+          likes:            @visualization.likes.count
```
How come you change this? Previous version is better because avoids exposing how visualization likes are counted.

gfiorav replied:
Well, I was kind of puzzled by this actually, the problem is that `likes_count` method doesn't exist for `Carto::Visualization`

juanignaciosl replied:
Ok, I see it must be a bug. I'm also puzzled indeed. Nevertheless, since `Visualization::Member` has that method I'd prefer implementing it at `Carto::Visualization`. It's trivial and avoids implementation leakage.

juanignaciosl replied:
`likes:            @visualization.likes.count` --> `likes:            @visualization.likes_count`

gfiorav replied:
changed one, missed the other :sweat_smile: now it should be ok

juanignaciosl commented on spec/requests/carto/editor/public/embeds_controller_spec.rb:
```diff
@@ -0,0 +1,81 @@
+      @visualization.privacy = Carto::Visualization::PRIVACY_PRIVATE
+      @visualization.save
+
+      get editor_visualization_public_embed_url(visualization_id: @visualization.id)
```
Check 403 error code.

juanignaciosl commented on spec/requests/carto/editor/public/embeds_controller_spec.rb:
```diff
@@ -0,0 +1,81 @@
+      @visualization.privacy = Carto::Visualization::PRIVACY_PROTECTED
+      @visualization.save
+
+      get editor_visualization_public_embed_url(visualization_id: @visualization.id)
```
Same, 403 error code check.

juanignaciosl commented on spec/requests/carto/editor/public/embeds_controller_spec.rb:
```diff
@@ -0,0 +1,81 @@
+      Carto::Visualization.any_instance.stubs(:password_valid?).with('manolo').returns false
+
+      post editor_visualization_public_embed_protected_url(visualization_id: @visualization.id, password: 'manolo')
+
```
Missing error code check as well.

juanignaciosl commented:
Tests miss some status error codes that must be added, but :+1: afterwards. Nice code and really nice and brave refactors, you rock! :clap: :dancer: :sparkles:

gfiorav commented:
Thanks for your kind words :_)

Tomorrow i'll test some old stuff again and merge!

gfiorav commented:
@juanignaciosl I added the return code check for the tests, plus returning those codes in the implementation and added a `likes_count` method for `Carto::Visualization`. See if you approve the changes please :pray:

juanignaciosl commented on app/models/carto/visualization.rb:
```diff
@@ -279,6 +283,10 @@ def can_be_cached?
   end
 
+  def likes_count
+    @likes_count ||= Carto::Like.where(subject: id).count
```
Try a better alternative, just returning `likes.count` should work and it's memoized, no need to do it manually.

gfiorav replied:
Ok! I was following the previous code's philosophy, but i agree it's better to do it like you propose.

juanignaciosl commented:
:+1: nice work!

gfiorav commented:
Thanks! When CI passes I'll deploy!
